_posts/2022-01-01-the-kubernetes-discovery-cache-blessing-and-curse.md

```diff
@@ -0,0 +1,101 @@
+
+```text
+GET https://<host>/api/v1/namespaces/<current namespace>/pods
+```
+
+You see that `kubectl` has to fill in the `<current namespace>` and query for `/pods` (and not for `/po` or `/pod`). It gets the `<current namespace>` through the `$KUBECONFIG` (which is usually stored at `~/.kube/config`), or falls back to `default`. It is also possible to query pods of all namespaces at once. The way `kubectl` resolves a request for `po` or `pod` to the final endpoint `/pods` is through a local cache stored at `~/.kube/cache/discovery/<host>/v1/serverresources.json`. In fact, there is a `serverresources.json` file for every group version of resources installed on the cluster. If you look at the entry for pods you will find something like
+
+```json
+{
+  "name": "pods",
+  "singularName": "",
+  "namespaced": true,
+  "kind": "Pod",
+  "verbs": [...],
+  "shortNames": [
+    "po"
+  ],
+  "categories": [
+    "all"
+  ]
+}
+```
+
+With this reference `kubectl` knows that a request for `pods`, `pod` (which is the `kind`), `po` (which is in the `shortNames` array) or `all` (which is in the `categories`) should result in the final request for `/pods`.
+
+`kubectl` creates the `serverresources.json` for every group version either if the requested kind is not present in any of the cached `serverresources.json` files, or if the cache is invalid. The cache invalidates itself [every 10 minutes](https://github.com/kubernetes/kubernetes/blob/0fb71846df9babb6012a7fce22e2533e9d795baa/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go#L253).
+
+That means in those cases `kubectl` has to make a request to every group version on the cluster to populate the cache again, which results in those 100s of `GET` requests described earlier, and those again trigger the client-side rate limiting. On large clusters with many CRDs `kubectl get` requests can often easily take up to a minute to run through all these requests plus pausing for the rate limiting. Thus it is advisable to not let your CRD count grow limitless. In fact, the scale targets for GA of custom resource definitions is set to 500 in the [Kubernetes enhancement repo](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/95-custom-resource-definitions#scale-targets-for-ga).
+
+So while the discovery cache is actually adding usability to Kubernetes, it also is the limiting factor for extending the platform with custom controllers and CRDs.
+
+### Scaling
+
+Especially the crossplane community has a vested interest in unlocking this limitation because crossplane's entire design philosophy is built upon the idea to create CRDs for every object in the real world and reconciling it through controllers. But it will also be important for other controllers introducing many CRDs like the [GCP Config Connector](https://github.com/GoogleCloudPlatform/k8s-config-connector) or the [Azure Service Operator](https://github.com/Azure/azure-service-operator).
```

Comment:
```suggestion
The crossplane community has an especially vested interest in unlocking this limitation, because crossplane's entire design philosophy is built upon the idea of creating CRDs for every object in the real world and reconciling it through controllers. But it will also be important for other controllers to introduce many CRDs like the [GCP Config Connector](https://github.com/GoogleCloudPlatform/k8s-config-connector) or the [Azure Service Operator](https://github.com/Azure/azure-service-operator).
```